### ISSUE TITLE
handling of polygons with overlapping holes

### ISSUE THREAD

**Issue by lightmare** [mapnik/mapnik#3953]
I stumbled upon this strange test file [test/data/json/poly-multihole.json](https://github.com/mapnik/test-data/blob/6c0b482802f81e5d87c70d7393b10e41f2d75398/json/poly-multihole.json?short_path=9e15754), that contains a Polygon with 3 rings:
 1. outer
 2. hole
 3. island in hole, or hole in hole, depending on how you look at it

**tl;dr: Do we need to support this kind of geometry?**

It just doesn't seem right. If I load it into QGIS, it renders a polygon with a hole and an island. But the island is *not inside*, it's an area twice subtracted, so clicking it won't select the feature.

In mapnik the hole in hole also renders as an island:
![poly](https://github.com/mapnik/test-data-visual/raw/bbeb5ec35c75fa8b5712f8215f132a19e84ea1fb/images/building-symbolizer-2-300-300-1.0-agg-reference.png).

In particular I want to know whether we need to support this behavior in BuildingSymbolizer. I've hit this when I tried to skip rendering back walls. Because the island has the winding order of an inner ring, I ended up with its back walls rendered and front walls skipped.

There's a similarly named shapefile [poly-multihole.shp](https://github.com/mapnik/test-data/blob/6c0b482802f81e5d87c70d7393b10e41f2d75398/shp/poly-multihole.shp) that contains a MultiPolygon with 2 Polygons: first polygon has a hole, second is the island inside that hole. Unlike the geojson version, the shapefile is unambiguous -- the island is a separate polygon and so is a part of the feature. But the file should rather be named `multipoly-hole.shp` to reflect its contents.

These two data files (json and shp) are used by visual tests named `multipolygon-two-holes-...` which is another case of confusing names. I can't figure out what are these supposed to test.
 - multipolygon with one hole -- `MultiPolygon(Polygon(outer,hole), Polygon(outer))` or
 - polygon with hole in hole -- `Polygon(outer, hole, island)`?

I hope it's the former :)

**Comment by flippmoke** [comment]
In general I believe that we follow the OGC Simple Feature model... I am not sure if this is an example test to see that we convert as properly internally to a multipolygon?

**Comment by lightmare** [comment]
Previously I only checked [PostGIS docs](https://postgis.net/docs/manual-2.4/using_postgis_dbmanagement.html#OGC_Validity) where there's no example of overlapping holes, so I wasn't sure whether it's valid or not.

Now reading through http://www.opengeospatial.org/standards/sfa the answer is clear:
> **6.1.11 Polygon, Triangle**
> e) The interior of every Polygon is a connected point set;
> f) The exterior of a Polygon with 1 or more holes is not connected. Each hole defines a connected component of the exterior.

An island in a hole violates e) because it's disconnected. If the island is instead treated as a hole, it violates f) because it doesn't define a component of the exterior.

**Comment by talaj** [comment]
You are right, @lightmare. I've used the polygon in the test as it was some complex polygon I found in the data, not realizing consequences. I think we should not assume this is a valid geometry, to avoid some logical traps and complexities.
